pkg/api/types.go

```diff
@@ -1733,6 +1733,44 @@ type NodeStatus struct {
 	Images []ContainerImage `json:"images,omitempty"`
 }
 
+// AvoidPods desribes pods that should avoid this node.
+type AvoidPods struct {
+	// Bounded-sized list of signatures of pods that should avoid this node, sorted
+	// in timestamp order from oldest to newest. Size of the slice is unspecified.
+	PreferAvoidPods []PreferAvoidPodsEntry `json:"preferAvoidPods,omitempty"`
+}
+
+// Describes a class of pods that should avoid this node.
+type PreferAvoidPodsEntry struct {
+	// The class of pods.
+	PodSignature PodSignature `json:"podSignature"`
+	// Time at which this entry was added to the list.
+	EvictionTime unversioned.Time `json:"evictionTime,omitempty"`
+	// (brief) reason for this entry was added to the list.
+	Reason string `json:"reason,omitempty"`
+	// Human readable message indicating why this entry was added to the list.
+	Message string `json:"message,omitempty"`
+}
+
+// Describes the class of pods that should avoid this node.
+// Exactly one field should be set.
+type PodSignature struct {
+	// Reference to controller whose pods should avoid this node.
+	PodController *PodController `json:"podController"`
+}
+
+// PodController contains enough information to let you inspect or modify the referred controller.
+type PodController struct {
+	// Kind of the controller.
+	Kind string `json:"kind,omitempty"`
+	// Namespace of the controller.
+	Namespace string `json:"namespace,omitempty"`
+	// Name of the controller.
+	Name string `json:"name,omitempty"`
+	// API version of the controller.
+	APIVersion string `json:"apiversion,omitempty"`
```

Comment:
Especially if we expect to be able to continue to look up the reference in future versions